File: polyglot/app_yaml_overwrites/py/src/app_yaml_overwrites/logger.py

import os
import logging
import json
from datetime import datetime
from typing import Optional, Dict, Any

class Logger:
    LEVELS = {
        'trace': logging.DEBUG - 5,
        'debug': logging.DEBUG,
        'info': logging.INFO,
        'warn': logging.WARNING,
        'error': logging.ERROR
    }

    def __init__(self, package_name: str, filename: str):
        self._package = package_name
        self._filename = filename
        self._level = os.environ.get('LOG_LEVEL', 'debug').lower()
        self._logger = logging.getLogger(f"{package_name}:{filename}")
        self._logger.setLevel(self.LEVELS.get(self._level, logging.DEBUG))
        
        # Ensure we have a handler for stdout
        if not self._logger.handlers:
            handler = logging.StreamHandler()
            self._logger.addHandler(handler)

    @classmethod
    def create(cls, package_name: str, filename: str) -> 'Logger':
        return cls(package_name, filename)

    def debug(self, message: str, **kwargs):
        self._log('debug', message, kwargs)

    def info(self, message: str, **kwargs):
        self._log('info', message, kwargs)
        
    def warn(self, message: str, **kwargs):
        self._log('warn', message, kwargs)

    def error(self, message: str, **kwargs):
        self._log('error', message, kwargs)

    def _log(self, level: str, message: str, data: Dict[str, Any]):
        if self._logger.isEnabledFor(self.LEVELS[level]):
            entry = {
                "timestamp": datetime.isoformat(datetime.now()),
                "level": level.upper(),
                "context": f"{self._package}:{self._filename}",
                "message": message,
                "data": data
            }
            # Standard Python logger format might interfere, so we just print JSON logic 
            # effectively mimicking the Node logger or letting standard logging handlers handle it.
            # For parity with Node, we use a structured dict passed to logging.
            self._logger.log(self.LEVELS[level], json.dumps(entry))

File: polyglot/app_yaml_overwrites/mjs/src/logger.ts

export class Logger {
    private package: string;
    private filename: string;
    private level: string;

    constructor(packageName: string, filename: string) {
        this.package = packageName;
        this.filename = filename;
        this.level = process.env.LOG_LEVEL?.toLowerCase() || 'debug';
    }

    static create(packageName: string, filename: string): Logger {
        return new Logger(packageName, filename);
    }

    debug(message: string, data?: Record<string, any>): void {
        if (this.shouldLog('debug')) {
            console.log(JSON.stringify({
                timestamp: new Date().toISOString(),
                level: 'DEBUG',
                context: `${this.package}:${this.filename}`,
                message,
                data
            }));
        }
    }

    info(message: string, data?: Record<string, any>): void {
        if (this.shouldLog('info')) {
            console.log(JSON.stringify({
                timestamp: new Date().toISOString(),
                level: 'INFO',
                context: `${this.package}:${this.filename}`,
                message,
                data
            }));
        }
    }

    warn(message: string, data?: Record<string, any>): void {
        if (this.shouldLog('warn')) {
            console.log(JSON.stringify({
                timestamp: new Date().toISOString(),
                level: 'WARN',
                context: `${this.package}:${this.filename}`,
                message,
                data
            }));
        }
    }

    error(message: string, data?: Record<string, any>): void {
        if (this.shouldLog('error')) {
            console.error(JSON.stringify({
                timestamp: new Date().toISOString(),
                level: 'ERROR',
                context: `${this.package}:${this.filename}`,
                message,
                data
            }));
        }
    }

    private shouldLog(level: string): boolean {
        const levels = ['trace', 'debug', 'info', 'warn', 'error'];
        const currentLevelIdx = levels.indexOf(this.level);
        const msgLevelIdx = levels.indexOf(level);
        // If invalid level setup, default to allowing INFO+
        if (currentLevelIdx === -1) return levels.indexOf('info') <= msgLevelIdx;
        return currentLevelIdx <= msgLevelIdx;
    }
}
